docs/resources/extract-ai.sh

```diff
@@ -6,26 +6,84 @@ BINARY_FOLDER=/var/mnt
 
 pushd $FOLDER
 
-total_copies=$(sort -u $BINARY_FOLDER/$OCP_RELEASE_LIST | wc -l)  # Required to keep track of the pull task vs total
-current_copy=1
-
-while read -r line;
-do
-  uri=$(echo "$line" | awk '{print$1}')
-  #tar=$(echo "$line" | awk '{print$2}')
-  podman image exists $uri
-  if [[ $? -eq 0 ]]; then
-      echo "Skipping existing image $tar"
-      echo "Copying ${uri} [${current_copy}/${total_copies}]"
+load_images() {
+
+  declare -A pids # Hash that include the images pulled along with their pids to be monitored by wait command
+  local max_bg=10 # Max number of simultaneous skopeo copies to container storage
+  local total_copies=$(sort -u $BINARY_FOLDER/$OCP_RELEASE_LIST | wc -l)  # Required to keep track of the pull task vs total
+  local current_copy=1
+
+  #remove duplicates
+  sort -u -o $OCP_RELEASE_LIST $OCP_RELEASE_LIST
+
+  while read -r line;
+  do
+    uri=$(echo "$line" | awk '{print$1}')
+    podman image exists $uri
+    if [[ $? -eq 0 ]]; then
+      echo "[INFO] Skipping existing image $tar"
+      echo "[INFO] Copying ${uri} [${current_copy}/${total_copies}]"
       current_copy=$((current_copy + 1))
       continue
+    fi
+    tar=$(basename ${uri/:/_})
+    tar --use-compress-program=pigz -xf ${tar}.tgz
+    if [ $? -eq 0 ]; then rm -f ${tar}.gz; fi
+    echo "[INFO] Copying ${uri} [${current_copy}/${total_copies}]"
+    skopeo copy dir://$(pwd)/${tar} containers-storage:${uri} -q &
+
+    pids[${uri}]=$! # Keeping track of the PID and container image in case the pull fails
+    max_bg=$((max_bg - 1)) # Batch size adapted 
+    current_copy=$((current_copy + 1)) 
+    if [[ $max_bg -eq 0 ]] || [[ $current_copy -gt $total_copies ]] # If the batch is done, then monitor the status of all pulls before moving to the next batch. If the last images are being pulled wait too.
+    then
+      for img in ${!pids[@]}; do
+        wait ${pids[$img]} # The way wait monitor for each background task (PID). If any error then copy the image in the failed array so it can be retried later
+        if [[ $? != 0 ]]; then
+          echo "[ERROR] Pull failed for container image: ${img} . Retrying later... "
+          failed_copies+=(${img}) # Failed, then add the image to be retrieved later
+        else
+          echo "[INFO] Removing folder for ${img}"
+          targz=$(basename ${img/:/_})
+          rm -rf ${targz}
+        fi
+      done
+      # Once the batch is processed, reset the new batch size and clear the processes hash for the next one
+      max_bg=10
+      pids=()
+    fi
+  done < ${BINARY_FOLDER}/${OCP_RELEASE_LIST}
+}
+
+retry_images() {
+  echo "[RETRYING]"
+  echo ""
+  local rv=0
+  for failed_copy in ${failed_copies[@]}; do
+      echo "[RETRY] Retrying failed image pull: ${failed_copy}"
+      tar=$(echo "$failed_copy" |  rev | cut -d "/" -f1 | rev | tr ":" "_")
```

Comment:
This should use the same parameter substitution / basename as the other tar name vars